specs/dwritecore/DWriteCore.md

```diff
@@ -1050,9 +1230,185 @@ int __cdecl wmain(int argc, WCHAR** argv)
+
+    HRESULT hr = dwriteFactory->TranslateColorGlyphRun(
+        baselineOrigin,
+        &glyphRun,
+        /*glyphRunDescription*/ nullptr,
+        desiredGlyphImageFormats,
+        paintFeatureLevel,
+        DWRITE_MEASURING_MODE_NATURAL,
+        reinterpret_cast<DWRITE_MATRIX const*>(&transform),
+        colorPaletteIndex,
+        &colorRunEnumerator
+    );
+
+    // Handle DWRITE_E_NOCOLOR, which is returned if the base glyph run does not
+    // have any color glyphs in the desired glyph image formats. In this case,
+    // just render the base glyph run as a monochrome glyph run.
+    if (hr == DWRITE_E_NOCOLOR)
+    {
+        deviceContext->DrawGlyphRun(
+            baselineOrigin,
+            &glyphRun,
+            foregroundBrush,
+            DWRITE_MEASURING_MODE_NATURAL
+        );
+        return S_OK;
+    }
+
+    // Any other error returned by TranslateColorGyphRun is a failure.
+    RETURN_IF_FAILED(hr);
+
+    // Solid color brush to be lazily created if needed.
+    wil::com_ptr<ID2D1SolidColorBrush> solidBrush;
+
+    // Use the returned enumerator object to iterate over the color glyph runs.
+    for (;;)
+    {
+        // Advanced to the first or next run.
+        BOOL haveRun;
+        RETURN_IF_FAILED(colorRunEnumerator->MoveNext(&haveRun));
+        if (!haveRun)
+            break;
+
+        // Get a pointer to the color glyph run structure.
+        DWRITE_COLOR_GLYPH_RUN1 const* colorGlyphRun;
+        RETURN_IF_FAILED(colorRunEnumerator->GetCurrentRun(&colorGlyphRun));
+
+        // Determine the brush to use for this color glyph run.
+        ID2D1Brush* runBrush;
+        if (colorGlyphRun->paletteIndex == 0xFFFF)
+        {
+            // Special palette index meaning use the text foreground brush.
+            runBrush = foregroundBrush;
+        }
+        else
+        {
+            // Use the specified color from the font's color palette.
+            if (solidBrush == nullptr)
+            {
+                // Lazily create the solid color brush with the specified color.
+                RETURN_IF_FAILED(deviceContext->CreateSolidColorBrush(colorGlyphRun->runColor, &solidBrush));
+            }
+            else
+            {
+                // We already created the brush, so just set its color.
+                solidBrush->SetColor(colorGlyphRun->runColor);
+            }
+
+            // use the solid color brush as the current run's brush.
+            runBrush = solidBrush.get();
+        }
+
+        // Branch depending on the run's glyph image format.
+        switch (colorGlyphRun->glyphImageFormat)
+        {
+        case DWRITE_GLYPH_IMAGE_FORMATS_NONE:
+            // do nothing
+            break;
+
+        case DWRITE_GLYPH_IMAGE_FORMATS_PNG:
+        case DWRITE_GLYPH_IMAGE_FORMATS_JPEG:
+        case DWRITE_GLYPH_IMAGE_FORMATS_TIFF:
+        case DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8:
+            deviceContext->DrawColorBitmapGlyphRun(
```

Comment:
I can't help but notice that all of these calls, more or less, take essentially the same arguments, and are essentially just passed the glyph run to render for the details.

Why have all these different calls then? Why not just have a "DrawSuperNewCoolGlyphRun", pass it the glyph run, and let it worry about the type?

Perhaps just my take knowing virtually nothing about DWrite or glyph rendering at the moment, but I'm struggling to understand the reasoning here, or what splitting these out buys us.

Reply:
Well, D2D could have a `DrawColorGlyphRun` method that takes a `DWRITE_COLOR_GLYPH_RUN1` and internally has the switch statement. In that case, the client would call `TranslateColorGlyphRun` and then forward each enumerated color glyph run to `DrawColorGlyphRun`. However, if you're going to go that far then it makes more sense to have a higher-level method that calls `TranslateColorGlyphRun` _for_ you, enumerates all the color runs, and then forwards each to the appropriate method. That's what the proposed `DrawGlyphRunWithColor` (or `DrawGlyphRunWithColorFontsEnabled`?) method does.

There are also cases where some clients want to handle different cases differently. E.g., XAML has different handling of monochrome vs. color glyph runs. Monochrome glyph runs are rendered to an alpha texture, while color runs are rendered to a BGRA32 texture.